cmake.md

```diff
@@ -0,0 +1,30 @@
+## CMake build system
+### Terminology
+
+The CMake system in symbiflow-arch-defs uses the following hierarchy:
+
+* ARCH - The architecture of FPGA.  This is used to capture architecture level parameters (e.g. tech libraries).
+* DEVICE_TYPE - A type of device.  This is defined as an ARCH and the arch.xml which defines the units with this device type.
+* DEVICE - A specific instance of a DEVICE_TYPE.  A device definition is a DEVICE_TYPE, a PACKAGE list, and IO map definitions for each package if needed.
+* BOARD - A specific DEVICE and PACKAGE instance, along with a PROG_TOOL/PROG_CMD.  A board is what is targeted by ADD_FPGA_TARGET.
+
+### Creating a new FPGA target
+
+Assuming that the board(s) you are interested in targetting is already defined, ADD_FPGA_TARGET is what will create targets.  See [ADD_FPGA_TARGET documentation](https://github.com/SymbiFlow/symbiflow-arch-defs/blob/master/make/devices.cmake#L559).
+
+If you want to target multiple boards, there is also ADD_FPGA_TARGET_BOARDS that will call ADD_FPGA_TARGET correctly for each board you request.  See [ADD_FPGA_TARGET_BOARDS](https://github.com/litghost/symbiflow-arch-defs/blob/re_enable_artix7_routing_tests/make/devices.cmake#L458).
+
+#### Note on ADD_FILE_TARGET and ADD_FPGA_BOARD
+
+All source files in the symbiflow-arch-defs are required to have a file target associated with them.  This is done via (ADD_FILE_TARGET)[https://github.com/SymbiFlow/symbiflow-arch-defs/blob/master/make/file_targets.cmake#L193].  By default ADD_FPGA_TARGET and ADD_FPGA_TARGET_BOARDS will both implicitly invoke ADD_FILE_TARGET for you.  This handles the common case where input sources are not generated files.  However, if you want to generate files, then you cannot use the implicity ADD_FILE_TARGET, and you will need to add ADD_FILE_TARGET calls for all inputs to ADD_FPGA_TARGET.
+
+For non-generated verilog, ADD_FILE_TARGET should be invoked like:
+```
+add_file_target(FILE <current source dir relative path> SCANNER_TYPE verilog)
+```
+
+For generated verilog, ADD_FILE_TARGET should be invoked like:
+```
+add_file_target(FILE <current source dir relative path> GENERATED)
+```
+Note that the input path must be current source dir relative, not absolute, and not relative to the current binary dir.  Also note that CMake cannot take depedendencies during build time, so for generated files you must supply the dependencies for the target.  See (APPEND_FILE_DEPENDENCY)[https://github.com/SymbiFlow/symbiflow-arch-defs/blob/master/make/file_targets.cmake#L79] for adding depedencies to other file targets.  It is generally suggested that all source forms within symbiflow-arch-defs are file targets, but it is required for verilog files because of how verilog handles relative include paths.
```

Comment:
I don't quite understand this last bit;
>  It is generally suggested that all source forms within symbiflow-arch-defs are file targets, but it is required for verilog files because of how verilog handles relative include paths.

Reply:
Verilog defines relative paths as relative to the location of the file, not to the location of the compiler.  As a result, if you use relative includes to include both a generated and non-generated file, things get exciting in out of source builds.  add_file_target solves this by copying non-generated files to the binary directory.